ui/src/SoundWave.js

```diff
@@ -0,0 +1,70 @@
+export class SoundWave {
+
+    constructor(canvas, stream) {
+        this.canvas = canvas;
+        this.stream = stream;
+        this.analyser = undefined;
+        this.stopId = undefined;
+        this.freqs = undefined;
+        this.ctx = undefined;
+        this.draw = this.draw.bind(this);
+    }
+
+    createSoundWave() {
+
+        this.ctx = this.canvas.getContext("2d");
+        const context = new (window.AudioContext || window.webkitAudioContext)();
+        this.analyser = context.createAnalyser();
+        const source = context.createMediaStreamSource(this.stream);
+        source.connect(this.analyser);
+
+        this.freqs = new Uint8Array(this.analyser.frequencyBinCount);
+
+        if(this.stream != undefined) {
+            this.stopId = requestAnimationFrame(this.draw);
+        }
+    }
+
+    stopSoundWave() {
+        this.stream = undefined;
+        cancelAnimationFrame(this.stopId);
+    }
+
+    draw() {
```

Comment:
nit: if possible, maybe break this up into a couple of helpers

Reply:
The length is about right here, IMO. Don't want it to get any longer though